pow/validator/src/main/java/org/ethereum/beacon/pow/validator/ValidatorRegistrationServiceImpl.java

```diff
@@ -0,0 +1,298 @@
+package org.ethereum.beacon.pow.validator;
+
+import org.ethereum.beacon.chain.observer.ObservableBeaconState;
+import org.ethereum.beacon.consensus.BlockTransition;
+import org.ethereum.beacon.consensus.SpecHelpers;
+import org.ethereum.beacon.consensus.StateTransition;
+import org.ethereum.beacon.consensus.hasher.ObjectHasher;
+import org.ethereum.beacon.consensus.hasher.SSZObjectHasher;
+import org.ethereum.beacon.consensus.transition.BeaconStateEx;
+import org.ethereum.beacon.consensus.transition.PerBlockTransition;
+import org.ethereum.beacon.consensus.transition.PerEpochTransition;
+import org.ethereum.beacon.core.BeaconState;
+import org.ethereum.beacon.core.operations.Deposit;
+import org.ethereum.beacon.core.operations.deposit.DepositInput;
+import org.ethereum.beacon.core.state.ValidatorRecord;
+import org.ethereum.beacon.core.types.BLSPubkey;
+import org.ethereum.beacon.core.types.BLSSignature;
+import org.ethereum.beacon.core.types.EpochNumber;
+import org.ethereum.beacon.core.types.Gwei;
+import org.ethereum.beacon.crypto.Hashes;
+import org.ethereum.beacon.db.source.SingleValueSource;
+import org.ethereum.beacon.pow.DepositContract;
+import org.ethereum.beacon.ssz.Serializer;
+import org.ethereum.beacon.validator.BeaconChainAttester;
+import org.ethereum.beacon.validator.BeaconChainProposer;
+import org.ethereum.beacon.validator.BeaconChainValidator;
+import org.ethereum.beacon.validator.ValidatorService;
+import org.ethereum.beacon.validator.attester.BeaconChainAttesterImpl;
+import org.ethereum.beacon.validator.crypto.MessageSigner;
+import org.ethereum.beacon.validator.proposer.BeaconChainProposerImpl;
+import org.reactivestreams.Publisher;
+import reactor.core.Disposable;
+import reactor.core.publisher.Flux;
+import tech.pegasys.artemis.ethereum.core.Address;
+import tech.pegasys.artemis.ethereum.core.Hash32;
+import tech.pegasys.artemis.util.bytes.Bytes8;
+import tech.pegasys.artemis.util.bytes.BytesValue;
+
+import javax.annotation.Nullable;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.function.Function;
+
+import static org.ethereum.beacon.core.spec.SignatureDomains.DEPOSIT;
+
+public class ValidatorRegistrationServiceImpl implements ValidatorRegistrationService {
```

Comment:
Maybe we don't need a complete service for that? I mean that it's not always necessary to run a full flow here. I think we can keep it as it is now and then break down to the components that would be really used.

Reply:
Not sure at this time. Will see it better when web3j or go added